sharness/test_cli_scores.t

```diff
@@ -0,0 +1,86 @@
+#!/bin/sh
+
+# Disable these lint rules globally:
+#   2034 = unused variable (used by sharness)
+#   2016 = parameter expansion in single quotes
+#   1004 = backslash-newline in single quotes
+# shellcheck disable=SC2034,SC2016,SC1004
+:
+
+test_description='tests for scripts/build_static_site.sh'
+
+export GIT_CONFIG_NOSYSTEM=1
+export GIT_ATTR_NOSYSTEM=1
+
+# shellcheck disable=SC1091
+. ./sharness.sh
+
+test_expect_success "environment and Node linking setup" '
+    toplevel="$(git -C "$(dirname "$0")" rev-parse --show-toplevel)" &&
+    snapshot_directory="${toplevel}/sharness/__snapshots__/" &&
+    SOURCECRED_DIRECTORY="${snapshot_directory}/example-github-load" &&
```

Comment:
You haven’t exported `SOURCECRED_DIRECTORY`, so it’s not visible to the
child processes. The children work against your actually-populated cache
locally, but fail with an empty cache on CI (or on my machine,
presumably, though I didn’t run this).

Immediate fix:

```sh
    SOURCECRED_DIRECTORY="${snapshot_directory}/example-github-load" &&
    export SOURCECRED_DIRECTORY &&
```

Long-term suggestion: In the top-level test config, make a temp
directory and export `` `${tmpdir}/ERROR_NO_SOURCECRED_DIRECTORY_SET` ``
to make tests more hermetic.

Also:

  - In your test, use `diff -u` rather than `diff -q` so that the output
    is printed on failure.

  - In your `run`, use something like

    ```sh
    run() (
      set -eu
      rm -f out err
      code=0
      node "${SOURCECRED_BIN}"/sourcecred.js "$@" >out 2>err || code=$?
      if [ "${code}" -ne 0 ]; then
        printf '%s failed with %d\n' "sourcecred $*"
        printf 'stdout:\n'
        cat out
        printf 'stderr:\n'
        cat err
      fi
      return "${code}"
    )
    ```

    so that you can see the output when it fails.

Reply:
Related to #1226

Reply:
Thanks for the suggestions! I adopted the all of them except the longer-term suggestion, which I didn't understand. Can you please elaborate on it? Are there any examples of a similar pattern already in the codebase?

Also, why should we separate defining SOURCECRED_DIRECTORY and exporting it into two separate statements? Couldn't we just as easily combine them into one?

Reply:
> I adopted the all of them except the longer-term suggestion, which I
> didn't understand. Can you please elaborate on it?

Sure.

When you run `config/test.js`, we invoke (e.g.) `yarn run -s sharness`
in a subprocess. That subprocess inherits your working environment,
including your `SOURCECRED_DIRECTORY` variable. If the Sharness task
doesn’t explicitly set the `SOURCECRED_DIRECTORY` variable itself, then
it will use your global cache, which is not hermetic: it may pass for
you but fail for everyone else, as happened here.

The simplest form of what I’m suggesting is to just add

```js
process.env["SOURCECRED_DIRECTORY"] = "/ERROR_SOURCECRED_DIRECTORY_NOT_SET";
```

to the top of `config/test.js`. Then, if the Sharness task fails to
explicitly set the `SOURCECRED_DIRECTORY` variable, it will inherit the
“dummy” value; any attempt to read from or write to this path, which is
a non-existent directory in a write-protected directory, will be much
more likely to fail.

In the case of the bug in the original version of this PR, I think that
the test would have just failed with

```
fatal: repository ID sourcecred/example-github not loaded
Try running `sourcecred load sourcecred/example-github` first.
```

when run locally, which is probably enough to point you in the right
direction. If the test had tried to run `sourcecred load …`, then the error message would have been

```
Contents of stderr:
    EACCES: permission denied, mkdir '/ERROR_SOURCECRED_DIRECTORY_NOT_SET'
```

which explicitly points to the problem in the test configuration.

(I say “simplest” because ideally we’d (a) change `yarn sharness`
similarly, to prevent skew between `yarn sharness` and `yarn test`, and
(b) use a directory name that is guaranteed to not exist by running
something like `tmpdir="$(mktemp -d)" && chmod 000 "${tmpdir}"` before
exporting `"${tmpdir}/ERROR_SOURCECRED_DIRECTORY_NOT_SET"`.)


> Are there any examples of a similar pattern already in the codebase?

Not exactly, no. But the general pattern is occasionally quite useful.

> Also, why should we separate defining SOURCECRED_DIRECTORY and
> exporting it into two separate statements? Couldn't we just as easily
> combine them into one?

Using `export NAME=word` is in general dangerous when `word` is not a
constant, because `export` returns success if the export itself succeeds
even if the expansion of `word` failed. For instance:

```shell
$ (set -e; FOO="$(false)"; printf 'got here\n'; export FOO); echo $?
1
$ (set -e; export FOO="$(false)"; printf 'got here\n'); echo $?
got here
0
```

It’s actually safe in this case, because the only non-constant portion
of the variable assignment is a parameter expansion, and failures of
parameter expansion (either unbound variables with `set -u`, or syntax
errors in the expansion form) _will_ cause the whole `export` to fail.
But this isn’t something that I really want to have to remember, hence
the blanket advice to separate `export`s from assignments whenever the
RHS of the assignment is non-constant.

That is: if you made this particular `export` also assign, I wouldn’t
object; the above habit is simply the reason that I wrote my suggestion
the way that I did.

There is a shellcheck lint for this:
<https://github.com/koalaman/shellcheck/wiki/SC2155>

(and the shellcheck lint _will_ only warn in the actually-dangerous
case, which is nice). I highly recommend integrating shellcheck-on-save
into your editor. It’ll catch all of those missing quotes!

Reply:
Thanks!